polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisPrincipal.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.auth;
+
+import java.security.Principal;
+import java.util.Map;
+import java.util.Set;
+import org.apache.polaris.core.entity.PrincipalEntity;
+import org.apache.polaris.immutables.PolarisImmutable;
+
+/** Represents a {@link Principal} in the Polaris system. */
+@PolarisImmutable
+public interface PolarisPrincipal extends Principal {
+
+  /**
+   * Creates a new instance of {@link PolarisPrincipal} from the given {@link PrincipalEntity} and
+   * roles.
+   *
+   * <p>The created principal will have the same ID and name as the {@link PrincipalEntity}, and its
+   * properties will be derived from the internal properties of the entity.
+   *
+   * @param principalEntity the principal entity representing the user or service
+   * @param roles the set of roles associated with the principal
+   */
+  static PolarisPrincipal of(PrincipalEntity principalEntity, Set<String> roles) {
+    return of(
+        principalEntity.getId(),
+        principalEntity.getName(),
+        principalEntity.getInternalPropertiesAsMap(),
+        roles);
+  }
+
+  /**
+   * Creates a new instance of {@link PolarisPrincipal} with the specified ID, name, roles, and
+   * properties.
+   *
+   * @param id the unique identifier of the principal
+   * @param name the name of the principal
+   * @param properties additional properties associated with the principal
+   * @param roles the set of roles associated with the principal
+   */
+  static PolarisPrincipal of(
+      long id, String name, Map<String, String> properties, Set<String> roles) {
+    return ImmutablePolarisPrincipal.builder()
+        .id(id)
+        .name(name)
+        .properties(properties)
+        .roles(roles)
+        .build();
+  }
+
+  /**
+   * Returns the unique identifier of the principal.
+   *
+   * <p>This identifier is used to uniquely identify the principal within a Polaris realm.
+   */
+  long getId();
```

Comment:
We differentiate between persisted (`PersistedPolarisPrincipal`) and non-persistent ones. I wonder whether this getter should be present at all here. It could at least return an `OptionalLong` to make it clear that only persisted principals have an ID. WDYT?

Reply:
The problem is that the Resolver requires the principal to have an id, since it resolves principals by id. Thus it seemed to me that it was legit to require from all principals to have at least a numeric identifier.

Reply:
As a follow-up task, we could attempt to modify the Resolver so that it makes principal lookups by name. WDYT @collado-mike ?

Reply:
I think we should be able to move past the Resolver needing to resolve the principal at all. This was largely present because the Resolver fetched the accessible PrincipalRoles based on grant records, but we have moved that requirement out so that we can rely on roles defined in the token instead. I think it's reasonable to change the Resolver to skip resolving the principal and roles entirely.